tracks/rust/exercises/bob/mentoring.md

```diff
@@ -0,0 +1,140 @@
+### Concepts
+
+- [bool](https://doc.rust-lang.org/std/primitive.bool.html)
+- [char](https://doc.rust-lang.org/std/char/index.html)
+- [match statement](https://doc.rust-lang.org/1.2.0/book/match.html)
+- [variables](https://doc.rust-lang.org/book/ch03-01-variables-and-mutability.html)
+- [closures](https://doc.rust-lang.org/book/ch13-01-closures.html)
+- [&str](https://doc.rust-lang.org/std/primitive.str.html)
+- [Iterators](https://doc.rust-lang.org/std/iter/trait.Iterator.html)
+- [Iterator Methods](https://doc.rust-lang.org/std/iter/trait.Iterator.html#provided-methods)
+- [constant](https://doc.rust-lang.org/std/keyword.const.html)
+
+### Reasonable Solutions
+
+A reasonable solution should:
+
+- Use constants
+- Use variables
+- Use `match` statements
+- Check if the input string is empty and return the "Sure" constant
+before doing any additional processing
+- Use simple methods from the standard library
+- Not use the regex crate
+- Not use separate functions
+
+#### Examples
+
+```rust
+const SURE: &str = "Sure.";
+const WHOA: &str = "Whoa, chill out!";
+const CALM: &str = "Calm down, I know what I'm doing!";
+const FINE: &str = "Fine. Be that way!";
+const WHATEVER: &str = "Whatever.";
+
+#[allow(clippy::match_bool)]
+pub fn reply(message: &str) -> &str {
+    let trimmed: &str = message.trim();
+
+    match trimmed.is_empty() {
+        true => FINE,
+        false => {
+            let is_question: bool = trimmed.ends_with("?");
+
+            let contains_alphabetic_characters: bool = trimmed.chars().any(|character| character.is_alphabetic());
+            let is_uppercase: bool = trimmed == trimmed.to_uppercase();
```

Comment:
This is pretty inefficient, since it's allocating a new `String` needlessly. I suggest making a `chars()` loop here to set both `contains_alphabetic_characters` and `is_uppercase` in one go.

Reply:
If you mean, `let is_yelled: bool = trimmed.chars().any(|character| character.is_alphabetic()) && trimmed == trimmed.to_uppercase();`, I prefer setting the separate variables to help explain what each piece is actually doing. I find it more readable. Now, I don't necessarily tell the students that they have to do that. If not, please elaborate more :)

Reply:
Yeah, being inefficient surely isn't a problem in exercises, but otoh doing it this way discards a bit of what a "low level" language can do, compared to a "higher level" language.

Again, if you write it this way in the mentoring notes, I'm afraid of mentors assuming this is the way it _has_ to be done. Maybe add a second, more efficient example solution, and add an inline comment referring to it?

Reply:
The best way to go about this is to, below this example block, add a section with similar wording to:

```markdown
For `is_yelled` , variations include using a `chars()` loop to ensure no extra `String` is allocated.
```

I suggest **not** adding inline comments, as some mentors see that as a mandatory comment in the solution and it's much clearer to keep the same structure everywhere (having variations below the solution and not inline, or both).